text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,165 @@
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget), the settlement stage
+   does not exist.
+
+The semantic conventions described below define how to model those stages in
+traces, how to propagate context, and how to enrich traces with attributes.
+
+### Context propagation
+
+Two layers of context propagation need to be distinguished for messaging
+workflows:
+
+1. The _creation context layer_ allows correlating the producer with the
+   consumers of a message, regardless of intermediary instrumentation. The
+   creation context is created by the producer and must be propagated to the
+   consumers. It must not be altered by intermediaries.
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. The _transport context layer_ allows correlating the producer and the
+   consumer with an intermediary. It also allows to correlate multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries
+   that are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+A producer MUST attach a creation context to each message. The creation context
+MUST be attached in a way so that it is not changed by intermediaries. A
+producer MAY propagate a transport context to an intermediary.  An
+intermediary MAY propagate a transport context to a consumer.
+
+### Span structure, names, and attributes
+
+#### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Receive"
+spans to correlate producer with consumer traces. A "Receive" span must link
+to the "Create" spans of all messages that are handled by the respective
+"Receive" operation. Depending on the use case, "Receive" spans can correlate
+with "Process" spans or other spans modelling processing operations.
```

Comment:
I would suggest that we state the exact details of this correlation in the specification, so tracing backends can rely on some consistent logic to process messaging traces.

My suggestion is to always "link" from the "process" span to the "receive" span. This is useful in cases of ambient context and broken traces, where "process" and "receive" can potentially reside in different traces or not have a parent-child relationship.

Reply:
I wonder if other people think that there are cases where it is beneficial to model "Process" spans as children of "Receive" spans. It seems to me there are such cases, if so we probably shouldn't prohibit that.

Reply:
Refer to https://github.com/open-telemetry/oteps/pull/192#discussion_r799879951 for further discussion.